modules/core/src/parallel_impl.cpp

```diff
@@ -580,8 +580,11 @@ void ThreadPool::run(const Range& range, const ParallelLoopBody& body, double ns
             pthread_mutex_unlock(&mutex);
 
             CV_LOG_VERBOSE(NULL, 5, "MainThread: wake worker threads...");
-            for (size_t i = 0; i < threads.size(); ++i)
+            size_t max_loop = std::min(static_cast<size_t>(range.size()), threads.size());
+            for (size_t i = 0; i < max_loop; ++i)
             {
+                if (job->current_task >= job->range.size())
```

Comment:
/cc @vrabaud Could you check if your version of TSAN is happy with that change? (no data-race with `current_task`)